### PR TITLE
Save screenshot to removable storage if needed and possible

### LINKED ISSUE #200: Long page screenshot enhancements: save to SD card
Depends on: #48

### PR DESCRIPTION
**Pull request by mTwTm**
Fixes #200

### CODE REVIEW

File: app/src/main/java/org/mozilla/focus/fragment/BrowserFragment.java [hunk 1]
```diff
@@ -785,7 +786,7 @@ private Bitmap getPageBitmap(WebView webView) {
     }
 
     private String saveBitmapToStorage(String fileName, Bitmap bitmap) throws IOException {
-        File folder = new File(Environment.getExternalStorageDirectory(), "Zerda");
+        File folder = StorageUtils.getTargetDirForSaveFile(getActivity(), "Zerda");
         if (!folder.exists() && !folder.mkdir()) {
```
Comment (walkingice) [inline app/src/main/java/org/mozilla/focus/fragment/BrowserFragment.java]: and there is `FileUtils.ensureDir(dir)` to use

File: app/src/main/java/org/mozilla/focus/utils/StorageUtils.java [hunk 2]
```diff
@@ -27,21 +50,14 @@
...
-        if (media == null) {
-            throw new NoRemovableStorageException("No removable media to use");
-        }
-
-        if (!Environment.MEDIA_MOUNTED.equals(Environment.getExternalStorageState(media))) {
-            throw new NoRemovableStorageException("No mounted-removable media to use");
-        }
+        final File media = testAndGetRemovableStorage(ctx);
 
         final File dir = new File(media, DOWNLOAD_DIR);
```
Comment (walkingice) [inline app/src/main/java/org/mozilla/focus/utils/StorageUtils.java]: We can refactor this method by removing this line, then rename to 
```diff
- public static File getTargetDirOnRemovableStorage(@NonNull Context ctx, String type)
+ public static File getAppMediaDirOnRemovableStorage(@NonNull Context ctx)
```

Then add this two functions
```java
public static File getTargetDirOnRemovableStorageForDownloads(@NonNull Context ctx, String mimeType) throws NoRemovableStorageException {
    File dir = getAppMediaDirOnRemovableStorage(ctx);
    if (dir == null) {
        return null;
    } else {
        return (MimeUtils.isImage(type)) ?  new File(dir, IMAGE_DIR) : new File(dir, OTHER_DIR);
    }
}

public static File getDirForSaveFile(@NonNull Context ctx, String folderName) {
    File dir = getAppMediaDirOnRemovableStorage(ctx);
    if (dir == null OR NoRemovableStorageException) {
        // user don't want to save to external dir, fallback to internal storage
        return new File(Envrionment....., folderName);
    } else {
        return new File(dir, folderName);
    }
}
```

File: app/src/main/java/org/mozilla/focus/utils/StorageUtils.java [hunk 3]
```diff
@@ -19,6 +19,29 @@
...
 
     /**
+     * Test if we have a removable storage and throw Exception if no is available and specify cause.
+     * Note that exception is thrown even if the user does not want to save to removable storage.
+     *
+     * @param ctx  Context
+     * @return root directory on removable storage.
+     * @throws NoRemovableStorageException if there is no removable storage to use.
+     */
+    private static File testAndGetRemovableStorage(@NonNull Context ctx)
```
[the remark this answers is not in the document]
Comment (mTwTm) [reply app/src/main/java/org/mozilla/focus/utils/StorageUtils.java]: Fixed

File: app/src/main/java/org/mozilla/focus/utils/StorageUtils.java [hunk 4]
```diff
@@ -19,6 +19,29 @@
...
+            throws NoRemovableStorageException {
+
+        final File media = getFirstRemovableMedia(ctx);
+        if (media == null) {
+            throw new NoRemovableStorageException("No removable media to use");
```
[the remark this answers is not in the document]
Comment (mTwTm) [reply app/src/main/java/org/mozilla/focus/utils/StorageUtils.java]: Per our discussion, this is reasonable.

File: app/src/main/java/org/mozilla/focus/utils/StorageUtils.java (hunk 2 above)
Comment (mTwTm) [reply app/src/main/java/org/mozilla/focus/utils/StorageUtils.java]: Per our discussion, this is reasonable.

File: app/src/main/java/org/mozilla/focus/fragment/BrowserFragment.java (hunk 1 above)
Comment (mTwTm) [reply app/src/main/java/org/mozilla/focus/fragment/BrowserFragment.java]: Fixed

**Pull request merged**